### PR TITLE
Added all information for test example

### PR DESCRIPTION
**Pull request by skwong25**
Logo image not printing - to …be de-bugged. More test cases to be added. Language internationalisation and currency features to be added. Sorry - I forgot to submit pull requests at each stage!

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -6,6 +6,7 @@
   "scripts": {
     "start": "node src/app.js",
     "langtest": "node src/lang-test.js",
+    "endtoendtest": "node src/app.js -o '../src/order_example.json' -n 'The Jade Dumpling' -a '3, Overlinks Road, Beswick' -d 'United Kingdom M11 3DD' -v GB1234567 -b '** THANKYOU **' -i images/close_enough.jpg -p '0161 456 0972' -w 'www.thejadedumpling.co.uk' -t '** GYOZA TO GYO **' ",
```
Comment (fractalwrench) [inline package.json]: consider calling this `e2etest` to make it easier to type?

Comment (fractalwrench) [inline package.json]: It's probably worth creating a bash script that contains the `node src/app.js -o....` command, as you'll presumably want to run multiple different tests to check that different command line args work correctly. https://www.taniarascia.com/how-to-create-and-use-bash-scripts/

File: TECHSPEC.md [hunk 2]
```diff
@@ -41,18 +41,18 @@ A Command Line Application for printing receipts to Terminal
...
+- Logo (--i, -image) - README guide should say for best results, this image should be square. 
 - Name (--n, -name) - Should not exceed 20 chars 
 - Address Line 1 (--a, -address1) - Number, Street, City should not exceed 53 chars. 
 - Addres Line 2 (--b, -address2) - Country & UK Postcode should not exceed 53 chars.  
 
 - VAT no. (--v, -vat) - 9 digits. Contains only digits, except first 2 elements are GB. 
 - Telephone number (Optional) (--t, -tel) : 10 or 11 digits. Contains only digits. 
 - Website (Optional) (--w, -website) : Max 48 chars. Must contain 'www.' and '.com' or '.co.uk'
-- Message (Optional) (--tm, -topmessage): Max 30 chars. 
+- Message (Optional) (--t, -topmessage): Max 30 chars.
```
Comment (fractalwrench) [inline TECHSPEC.md:52]: One thing we didn't consider is the name of the file that is output - we should probably provide a command line arg to add this.

File: src/order_example.json [hunk 3]
```diff
@@ -0,0 +1,20 @@
+{
+    "items": { 
+            "1": {
```
Comment (fractalwrench) [inline src/order_example.json:3]: It would be more conventional to use an array here

File: src/order_example.json [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+                "value": 3.99
+            },
+            "2": {
+                "id": 2, 
+                "description": "dipping sauce",
+                "quantity": 1,
+                "value": 1.99
+            }
+        },        
+    "payment" : {
```
Comment (fractalwrench) [inline src/order_example.json:16]: perhaps we should just put the total in the JSON body supplied to us, rather than calculating it? I know I said otherwise, but I don't think it's massively important..

File: src/business.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
+// format the JSON order information into strings
+
+// const { createCanvas, loadImage, Image } = require('canvas')
+
+class BusinessInfo {
```
Comment (fractalwrench) [inline src/business.js:5]: It'd be good to add a JSDoc comment at the top of each of these classes which briefly explains what each one does.

File: src/business.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        if (this.options.phone) {
+        return `Tel: ${this.options.phone}\r\n`;  
+        } else {
+        return " ";
+        }
+    }
+
+    get website() {
+        if (this.options.website) {
+        return `Website: ${this.options.website}\r\n`;
```
Comment (fractalwrench) [inline src/business.js:32]: It would be good to run ESLint on the whole project as it will catch style violations such as these

File: src/business.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        return `${this.options.address1}\r\n${this.options.address2}\r\n`
+    }
+
+    get vat()  {
+        return `VAT NO: ${this.options.vat}\r\n`;
+    }
+
+    get phone() {
+        if (this.options.phone) {
+        return `Tel: ${this.options.phone}\r\n`;
```
Comment (fractalwrench) [inline src/business.js:24]: Is the carriage return `\r` required?

File: src/business.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+class BusinessInfo {
+    constructor(options) {
+        this.options = options;
+        this.sizeOf = require('image-size');
+
+        this.determineScale = this.determineScale.bind(this); 
+        // TODO: check if we need to bind determineScale method actually? 
+    }
+    
+    get address() {
```
Comment (fractalwrench) [inline src/business.js:14]: It would be nice to add some unit tests to verify what happens on each of these functions when passing in `null/undefined` - I think the fallback currently works well for some of them, but not others

File: src/items.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
+// format the JSON order information into strings
+
+// const { createCanvas, loadImage, Image } = require('canvas')
+
+class Items {
```
Comment (fractalwrench) [inline src/items.js:5]: Consider renaming this to `OrderInformation`?

File: src/business.js [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+      
+       if (widthDiff >= heightDiff) { 
+         return logoWidth/oWidth;  // 448/224 = 2 or 448/896=0.5
+       } else {
+         return logoHeight/oHeight; 
+       }
+    }
+
+    // calculates new dimensions for logo image 
+    getNewDimensions (url) {
```
Comment (fractalwrench) [inline src/business.js:67]: This muddies the water about responsibilities - a cleaner approach that avoids the need for requiring the `image-size` module entirely would be to extract this function out of this class, and put it with the drawing functions.

File: src/items.js [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    constructor(order) {
+        this.canvas = require('canvas');
```
Comment (fractalwrench) [inline src/items.js:7]: This can be removed

File: src/items.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        });
+        return grandTotal; 
+    }
+
+    // access this.order.items - an array 
+    // i want to make an array of the descriptions
+    makeArray(key) {
+        let arr = [];
+        for (let item in this.order.items) {
+            arr.push(this.order.items[item][key]);
```
Comment (fractalwrench) [inline src/items.js:35]: Calling `this.order.items[item]` seems a bit strange - is it not possible to just use `item`?

File: src/payment.js [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+        return this.payment.cash;
+    }
+
+    formatText() {
+        const currency = "GBP" // we can swap this out for currency chosen 
+        if (this.payment.type === "cash") {
+            const paid = this.payment.paid.toFixed(2)
+            const line1 = `PAID in ${currency}:   ${paid}\r\n`; 
+            const change = this.payment.paid - this.grandTotal; 
+            const line2 = `CHANGE in ${currency}: ${change.toFixed(2)}`
```
Comment (fractalwrench) [inline src/payment.js:23]: I think it might be worth using a library specifically designed to work with currency, such as this: https://currency.js.org/

As this avoids the pitfalls of floating point precision

File: src/app.js [hunk 14]
```diff
@@ -1,189 +1,254 @@
...
-  let heightDiff = diff(logoHeight, oHeight);
 
-  if (widthDiff <= heightDiff) {
-    return oWidth/logoWidth;  
-  } else {
-    return oHeight/logoHeight; 
-  }
+
+// sets heights for elements 
+const margin = 60;
```
Comment (fractalwrench) [inline src/app.js:51]: I think all the drawing code should go in a separate file

File: src/app.js [hunk 15]
```diff
@@ -1,189 +1,254 @@
...
+
+const maxSpacing = 40;
+const midSpacing = 30;
+const minSpacing = 20;
+const linebreak = 20; 
+const dottedLineBreak = 20;
```
Comment (fractalwrench) [inline src/app.js:57]: nice use of constants, this is much more readable 👍

File: src/app.js [hunk 16]
```diff
@@ -1,189 +1,254 @@
...
+let centrepoint = canvasWidth/2 
+
+
+
+// functions
+// TODO: these can be extracted into another module to clean up the main app.js
+const generateTimeStamp = () => {
+  const today = new Date();
+  const date = today.getFullYear()+"-"+(today.getMonth()+1)+"-"+today.getDate();
+  const time = today.getHours() + ":" + today.getMinutes() + ":" + today.getSeconds();
```
Comment (fractalwrench) [inline src/app.js:83]: `toLocaleDateString()` looks like it might be sufficient for this task: https://stackoverflow.com/a/34015511

There are a lot of different rules for formatting date + time across various countries, so it's better to go with the default Locale behaviour whenever there isn't an explicit requirement to format a date in a particular way.

File: src/app.js [hunk 17]
```diff
@@ -1,189 +1,254 @@
...
+  return date + " " + time; 
 }
-let scale = determineScale(); 
-const newWidth = oWidth * scale 
-console.log("newWidth: " + newWidth);
-const newHeight = oHeight * scale 
-console.log("newHeight: " + newHeight);
 
+const drawLogo = async function (url) { 
+  await canvas.loadImage(url)
```
Comment (fractalwrench) [inline src/app.js:88]: I've confirmed that this gets called last of all the drawing functions. The image has a height of 780 which is why it's not displayed on the image - so the fix will be to make sure this gets called first

File: src/app.js [hunk 18]
```diff
@@ -1,189 +1,254 @@
...
-    context.drawImage(image, 284, 120, newWidth, newHeight) 
-    const buffer = canvas.toBuffer('image/png')
-    fs.writeFileSync('./image.png', buffer)
-})
-.catch(error => {console.log(error)});
-// TODO: check if we really need to 'load' the image again 
+
+
+
+// validates CLI arguments and reassigns reformatted value
```
Comment (fractalwrench) [inline src/app.js:142]: validation should really happen before drawing takes place

File: src/app.js [hunk 19]
```diff
@@ -1,189 +1,254 @@
...
+// optional business info 
+drawOptInfo(); 
+
+// optional top message
 drawTopMessage(); 
 
-// draw dotted line break
-const dotBreak = '-----------------------------' // +  SPACE
+// dotted line break
+drawDottedLine()
```
Comment (fractalwrench) [inline src/app.js:204]: These functions are really great - it'd be good to encapsulate other draw statements in a similar way

**Review by fractalwrench — APPROVED** [review]
This looks really good 🥇 

I left some comments but I think the only remaining thing to do is to setup CI, and to add a short readme of 2-3 paragraphs explaining what the project is and how to run it, along with some images. The rest of the comments can be entered into Trello & addressed at your leisure. Well done!

File: src/app.js (hunk 18 above)
Comment (skwong25) [reply src/app.js:142]: It does. It goes after declaration of JS functions but before the 'draw' functions are called.
